notebooks/MAST/HSC/CASJOBS/HCV_CASJOBS/HCV_casjobs_demo.ipynb

```diff
@@ -0,0 +1,995 @@
+    "# get an array that has only one row per object\n",
+    "mval, uindex = np.unique(tab['MatchID'], return_index=True)\n",
+    "utab = tab[uindex]\n",
+    "print(f\"{len(utab)} unique MatchIDs in table\")\n",
+    "\n",
+    "tab"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "An `ExpertClass` value of 1 indicates that the object is confidently confirmed to be a variable; 2 means that the measurements do not have apparent problems and so the object is likely to be variable (usually the variability is too small to be obvious in the image); 4 means that the variability is likely to be the result of artifacts in the image (e.g., residual cosmic rays or diffraction spikes from nearby bright stars).\n",
+    "\n",
+    "Compare the distributions for single-filter variable candidates (SFVC, `AutoClass`=1) and multi-filter variable candidates (MFVC, `AutoClass`=2).  The fraction of artifacts is lower in the MFVC sample."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sfcount = np.bincount(utab['ExpertClass'][utab['AutoClass'] == 1])\n",
+    "mfcount = np.bincount(utab['ExpertClass'][utab['AutoClass'] == 2])\n",
+    "sfrat = sfcount/sfcount.sum()\n",
+    "mfrat = mfcount/mfcount.sum()\n",
+    "\n",
+    "print(\"Type Variable Likely Artifact Total\")\n",
+    "print(\"SFVC {:8d} {:6d} {:8d} {:5d} counts\".format(sfcount[1], sfcount[2], sfcount[4], sfcount.sum()))\n",
+    "print(\"MFVC {:8d} {:6d} {:8d} {:5d} counts\".format(mfcount[1], mfcount[2], mfcount[4], mfcount.sum()))\n",
+    "print(\"SFVC {:8.3f} {:6.3f} {:8.3f} {:5.3f} fraction\".format(sfrat[1], sfrat[2], sfrat[4], sfrat.sum()))\n",
+    "print(\"MFVC {:8.3f} {:6.3f} {:8.3f} {:5.3f} fraction\".format(mfrat[1], mfrat[2], mfrat[4], mfrat.sum()))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Plot the MAD variability index distribution with expert classifications <a name=\"mad_expert\"></a>\n",
+    "\n",
+    "Note that only the filters identified as variable (`FilterDetFlag` > 0) are included here.\n",
+    "\n",
+    "This version of the plot shows the distributions for the various `ExpertClass` values along with, for comparison, the distribution for all objects in gray (which is identical in each panel).  Most objects are classified as confident or likely variables.  Objects with lower MAD values (indicating a lower amplitude of variability) are less likely to be identified as confident variables because low-level variability is more difficult to confirm via visual examination."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "w = np.where(tab['FilterDetFlag'] > 0)[0]\n",
+    "mad = tab['MAD'][w]\n",
+    "e = tab['ExpertClass'][w]\n",
```

Comment:
Here, and I *think* literally everywhere else in this notebook where `np.where` is used, you don't need the where.  (this is likely a muscle-memory thing from IDL users, where `where` *is* necessary.)  Instead something like:

```suggestion
    "msk = (tab['FilterDetFlag'] > 0)[0]\n",
    "mad = tab['MAD'][msk]\n",
    "e = tab['ExpertClass'][msk]\n",
```
(although I'm not entirely sure in this case if the `[0]` is needed or not)

Similar for the below throughout this notebook whenever `np.where` is used.

Reply:
Actually I disagree with this.  The above code with `msk = (...)[0]` is actually a bug -- it returns the value of `tab['FilterDetFlag'][0]>0`, i.e. a scalar value with the boolean for just the first row.

And even if you fix that error by removing the `[0]`, the version without `np.where` will give equivalent results but leads to extra computation.  In a corrected version of the above example:
```
msk = tab['FilterDetFlag']>0
mad = tab['MAD'][msk]
e = tab['ExpertClass'][msk]
```
What that does is create a boolean array with `len(tab)` elements.  Then the indexing operations `[msk]` scans that array twice to find the `True` values in `msk` and uses that list of indices.  In contrast, my version using `np.where()`:
```
w = np.where(tab['FilterDetFlag']>0)[0]
mad = tab['MAD'][w]
e = tab['ExpertClass'][w]
```
... the list of indices `w` is generated only once, saving the repeated operation of scanning the boolean array for the two assignment statements.  If the table is large and the number of true elements is small, this can be a signficant savings.

My preference is to use `np.where` when the boolean is going to be used to index more than one array, and to simply use the boolean expression otherwise.  In the second case I would use just
`mad = tab['MAD'][tab['FilterDetFlag']>0]`
I use that style when it is appropriate.  Elsewhere we should keep the `np.where` version.

And yes, I am an old-time IDL user! 😜 But in this case I have a good reason for using `np.where`.

Reply:
In [HCV_API_demo.ipynb](https://github.com/trenton3983/notebooks/blob/master/notebooks/MAST/HSC/CASJOBS/HCV_API/HCV_API_demo.ipynb) and [HCV_casjobs_demo.ipynb](https://github.com/trenton3983/notebooks/blob/master/notebooks/MAST/HSC/CASJOBS/HCV_CASJOBS/HCV_casjobs_demo.ipynb), all instances of `np.where` have been removed. I made a similar comment to Erik in Slack about the `len` of the objects.

The original plots used `pylab.`, which I drop-in replaced with `plt.` from `import matplotlib.pyplot as plt`. At the time, I didn't refactor the plots. However, in the two mentioned notebooks, all the plotting has been refactored to use the object oriented API. The instances of the plots that required the use of `len(w)` (for labels) have been removed.

The relevant plots where in sections Plot HCV MAD variability index versus magnitude in F475W and Plot variables in color-magnitude diagram. 

Refactoring the plots significantly reduced the required code and other stored variables.

Additionally, in almost all of the cases with something like

```python
# create Boolean masks
w475 = tab['Filter'] == 'ACS_F475W'
w814 = tab['Filter'] == 'ACS_F814W'

# the only key needed to do the join is MatchID, but we include other common columns
# so that join includes only one copy of them
jtab = join(tab[w475], tab[w814],
            keys=['MatchID', 'GroupID', 'SubGroupID', 'RA', 'Dec', 'AutoClass', 'ExpertClass'],
            table_names=['f475', 'f814'])
print(len(jtab), "matched F475W+F814W objects")
jtab[jtab['AutoClass'] > 0]
```

I would do 

```python
# the only key needed to do the join is MatchID, but we include other common columns
# so that join includes only one copy of them
jtab = join(tab[tab['Filter'] == 'ACS_F475W'], tab[tab['Filter'] == 'ACS_F814W'],
            keys=['MatchID', 'GroupID', 'SubGroupID', 'RA', 'Dec', 'AutoClass', 'ExpertClass'],
            table_names=['f475', 'f814'])
print(len(jtab), "matched F475W+F814W objects")
jtab[jtab['AutoClass'] > 0]
```

Because `w475` and `w814` are never used again.

Reply:
Cases of code like `w = np.where(tab['FilterDetFlag']>0)[0]` are unchanged.

Reply:
In general using `np.where` is slower the just using the mask https://stackoverflow.com/q/67468493/7758804

![image](https://user-images.githubusercontent.com/12875240/181641179-578ab71e-fad2-4f9b-9137-3d00e5068790.png)

Reply:
With an example from the notebook

![image](https://user-images.githubusercontent.com/12875240/181643010-a96d0b5e-426d-4a82-83ef-0dd26e1be2b7.png)

Reply:
Yes, but that is exactly the wrong thing to be timing.  My use case is where you set `w = np.where(bool)` and then use `w` repeatedly. And that can be much faster than the boolean.  Here is an example.  First the equivalent of your test (and the one in stackoverflow), but using a sparser boolean array:
```
y = np.random.random(1000)
mask = y<0.01

%timeit y[mask]
%timeit y[np.where(mask)]
```
```
690 ns ± 20.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
1.21 µs ± 25.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
That's 1.75 times slower for `np.where`.  But now assign the `np,where` value to a variable rather than recomputing it every time (which is exactly the point):
```
w = np.where(mask)

%timeit y[mask]
%timeit y[w]
```
```
676 ns ± 17.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
193 ns ± 1.61 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```
That's 3.75 times faster.  And the factor can get much bigger with larger arrays.  For example:
```
y = np.random.random(100000)
mask = y<0.001
w = np.where(mask)
%timeit y[mask]
%timeit y[w]
```
```
15.1 µs ± 375 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
276 ns ± 3.86 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
That is 55 times faster!  My view is that using `np.where` routinely is a good idea.  To me the modified version of the code is much more difficult to read than the version I had with `w475` and `w814`.

Reply:
Also note that the penalty for calling `np.where` for even a single use array decreases significantly for large arrays.
```
y = np.random.random(100000)
mask = y<0.001
w = np.where(mask)
%timeit y[mask]
%timeit y[np.where(mask)]
%timeit y[w]
```
```
15.8 µs ± 785 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
16.9 µs ± 548 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
296 ns ± 5.35 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
So `np.where` rules IMO!

Reply:
In the cases where it's used more than once, like **Plot the fraction of artifacts as a function of MAD variability index**, `np.where` is faster. But are we doing `np.where(all_count > 0)` instead of `np.where(all_count > 0)[0]`, because the extra `[0]` doesn't seem to matter. That's my understanding from your initial comment.

In the case of `np.where(all_count[::-1].cumsum() > 40)[0][0]`, where you're getting the first value, isn't that what `np.argmax` is for `np.argmax(all_count[::-1].cumsum() > 40)`?

Reply:
I see

```
wnz = np.where(all_count > 0)[0]
nnz = len(wnz)
```

isn't the same as 

```
wnz = np.where(all_count > 0)
nnz = len(wnz)
```

Reply:
`np.where()` returns a tuple of values, with one array for each dimension.  You can use the return value directly to index a N-dimensional array.  When I know the array I'm using is 1-D, I prefer to just pull out the index array for that dimension by doing `w = np.where(boolean)[0]`.  That gets an array of all the indices for axis=0 (which is the only axis for a 1-D array).

For me the benefit of doing it that way is that `len(w)` is then the number of indices selected.  If I just use `w = np.where(boolean)` then I have to remember to use `len(w[0])`.  If I forget that, the value `len(w)` is always just 1 (the number of dimensions of the array).

The `cumsum` thing is a little tricky but is not the equivalent of `argmax`.  It's maybe easier to understand in the first version of with (without the `[::-1]` to reverse the array):
```
iz = np.where(all_count.cumsum() > 10)[0][0]
```
Breaking it down, `all_count.cumsum()` is a cumulative sum of the counts in the histogram bins.  `np.where(all_count.cumsum() > 10)[0]` is then the list of all the indices where the cumulative sum exceeds 10.  Indexing it by zero (the last `[0]`)  selects the first one of those indices.  So this expression figures out the number of bins you have to add up on the left edge of the histogram (starting with bin zero) to get > 10 total counts.  The other version that you quote reverses the array and does the same calculation from the right edge of the histogram (starting from `len(all_count)` and working down).

This whole thing is a little clumsy.  People have asked for `numpy` to add a shortcut function that returns the first element in an array where a boolean is true (or maybe the first element that is greater than some value, less than some value, etc.)  Instead of `argmax`, it would be something like `argfirst`.  But as far as I know, there is still no such function.  This slightly messy code is my way of working around that missing function without resorting to a loop (which is something I never do except under extreme duress.)

Reply:
By the way, you say up above: "The instances of the plots that required the use of `len(w)` (for labels) have been removed."  That's an example of the reason I like to extract the first element of the `where` tuple.  What do you do in those instances to preserve the counts in the labels?

Reply:
```python
# define plot parameter lists
auto_class = np.unique(jtab['AutoClass'])
markers = ['x', 'o', 'o']
colors = ['silver', 'blue', 'tab:cyan']
labels = ['non-', 'single-filter ', 'multi-filter ']

fig, ax = plt.subplots(figsize=(15, 10))
for ac, marker, color, label in zip(auto_class, markers, colors, labels):
    data = jtab[jtab['AutoClass'] == ac]
    ax.plot('MeanCorrMag_f475', 'MAD_f475', marker, markersize=4, color=color,
            label=f'{len(data):,} {label}variable candidates', data=data)

ax.set(xlabel='ACS_F475W [mag]', ylabel='ACS_F475W MAD [mag]')
ax.legend(loc='best', title=f'{len(jtab)} HSC measurements near {target}')
```

`data = jtab[jtab['AutoClass'] == ac]`

so `len(data)` is used

Reply:
That code could easily be consolidated to create a single figure with two subplots

```python
# define plot parameter lists
auto_class = np.unique(jtab['AutoClass'])
markers = ['x', 'o', 'o']
colors = ['silver', 'blue', 'tab:cyan']
labels = ['non-', 'single-filter ', 'multi-filter ']

# add a new column to jtab
jtab['MCMf475-MCMf814'] = jtab['MeanCorrMag_f475'] - jtab['MeanCorrMag_f814']

fig, ax = plt.subplots(2, 1, figsize=(15, 20))
for ac, marker, color, label in zip(auto_class, markers, colors, labels):
    data = jtab[jtab['AutoClass'] == ac]
    ax[0].plot('MeanCorrMag_f475', 'MAD_f475', marker, markersize=4, color=color,
            label=f'{len(data):,} {label}variable candidates', data=data)
    ax[1].plot('MCMf475-MCMf814', 'MeanCorrMag_f475', marker, markersize=4, color=color,
            label=f'{len(data):,} {label}variable candidates', data=data)

ax[0].set(xlabel='ACS_F475W [mag]', ylabel='ACS_F475W MAD [mag]')
ax[1].invert_yaxis()
ax[1].set(xlim=(-1.1, 4), xlabel='ACS F475W - F814W [mag]', ylabel='ACS F475W [mag]')

ax[0].legend(loc='best', title=f'{len(jtab)} HSC measurements near {target}')
ax[1].legend(loc='best', title=f'{len(jtab)} HSC measurements near {target}')
```

Reply:
That looks good to me -- I don't know that the switch to the object-oriented API is the key change, but I like the loop with `zip`, which definitely cleans up the code.